### PR TITLE
#16 [Shaz] Get all slots

### PR DESCRIPTION
**Pull request by shaz-ahammed**

### CODE REVIEW

File: controllers/slot_controller.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// @Failure 500 "Internal server error"
+func (sc SlotController) HandleGetAllSlots(ctx *gin.Context) {
+	page, _ := strconv.Atoi(ctx.Query("page"))
+	if page < 0 {
+		ctx.JSON(400, gin.H{"msg": "Page number must be greater than 0"})
+		return
+	}
+	status := ctx.Query("status")
+	if len(status) <= 0 {
+		status = ALL_SLOTS
```
Comment (vraravam) [inline controllers/slot_controller.go]: dont go with a string - which can be anything. For our purposes, we only want to filter `isAvailable` true/false (boolean ie only 2 possible values)

Comment (vraravam) [reply controllers/slot_controller.go]: Also, since you have implemented a filter, where are the appropriate tests?

File: models/factory/slot_factory.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+func GenerateRandomTimestamp() time.Time {
+	format := "2006-01-02 15:04:05"
+	timeObject, _ := time.Parse(format, faker.Timestamp())
+	return timeObject
+}
+
+func GenerateRandomStatus() string {
+	statusList := []string{"available", "booked", "reserved"}
+	return statusList[rand.Intn(len(statusList))]
+
```
Comment (vraravam) [inline models/factory/slot_factory.go]: EMPTY LINE!!!!!

File: repositories/slot_repository.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+type ISlotRepository interface {
+	RetrieveAllSlots(int, string) ([]models.Slot, error)
+}
+
+func (sr *ServiceRepository) RetrieveAllSlots(page int, status string) ([]models.Slot, error) {
+	var slots []models.Slot
+	
+	query := sr.db.Offset(page * DEFAULT_PAGE_SIZE).Limit(DEFAULT_PAGE_SIZE)
+	if status != "ALL" {
```
Comment (vraravam) [inline repositories/slot_repository.go]: Dont use a string - we don't need to support more than 2 possible values, which can be done easier via a boolean

**Review by vraravam — CHANGES_REQUESTED** [review]

File: controllers/slot_controller.go (hunk 1 above)
Comment (vraravam) [reply controllers/slot_controller.go]: also, availability check/filter can be combined with a time-period as well (just a suggestion for future enhancement)

Comment (shaz-ahammed) [reply controllers/slot_controller.go]: filter implementation tests come at the service layer or in our case repo layer

Comment (vraravam) [reply controllers/slot_controller.go]: you will be calling the service/repo with different arguments based on logic in the controller. so, that specific logic has to be tested only in the controller layer ie whether valueA vs valueB gets passed to service/repo. Isn't that right?

File: controllers/gate_controller.go [hunk 4]
```diff
@@ -36,16 +36,16 @@ func NewGateController(repository repositories.IGateRepository) *GateController
...
-	if err != nil || page < 0 {
+	// TODO: Convert to using a pagination library to handle this and other edge cases
+	page, _ := strconv.Atoi(ctx.Query("page"))
+	if page < 0 {
 		page = 0
 	}
-	floor, err := strconv.Atoi(floorStr)
-	if err != nil || floor < 0 {
+	floor, _ := strconv.Atoi(ctx.Query("floor"))
+	if floor < 0 {
```
Comment (vraravam) [inline controllers/gate_controller.go]: Need tests for this logic in the controller.

File: controllers/gate_controller.go [hunk 5]
```diff
@@ -36,16 +36,16 @@ func NewGateController(repository repositories.IGateRepository) *GateController
 func (gc *GateController) HandleGetAllGates(ctx *gin.Context) {
 	log.Debug().Msg("Getting list of gates")
 
-	pageStr := ctx.Query("page")
-	floorStr := ctx.Query("floor")
-	page, err := strconv.Atoi(pageStr)
-	if err != nil || page < 0 {
+	// TODO: Convert to using a pagination library to handle this and other edge cases
+	page, _ := strconv.Atoi(ctx.Query("page"))
+	if page < 0 {
```
Comment (vraravam) [inline controllers/gate_controller.go:43]: Need tests for this logic in the controller.

File: controllers/slot_controller.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+// @Tags slot
+// @Produce  json
+// @Param   page        query    int     false        "Page number (default = 1)"
+// @Param   is_available        query    boolean     false        "Availablity of the slot. [true, false] (default = true)"
+// @Success 200  "ok"
+// @Failure 500 "Internal server error"
+func (sc SlotController) HandleGetAllSlots(ctx *gin.Context) {
+	// TODO: Convert to using a pagination library to handle this and other edge cases
+	page, _ := strconv.Atoi(ctx.Query("page"))
+	if page < 0 {
```
Comment (vraravam) [inline controllers/slot_controller.go:36]: Need tests for this logic in the controller.

File: models/factory/slot_factory.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+func GenerateRandomTimestamp() time.Time {
+	format := "2006-01-02 15:04:05"
+	timeObject, _ := time.Parse(format, faker.Timestamp())
+	return timeObject
+}
+
+func GenerateRandomStatus() string {
+	statusList := []string{"available", "booked", "reserved"}
+	arr, _ := faker.RandomInt(0, 2, 3)
```
Comment (vraravam) [inline models/factory/slot_factory.go]: the indices are either `1,2,3` or `0,1,2`

File: repositories/slot_repository.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+type ISlotRepository interface {
+	RetrieveAllSlots(int, bool) ([]models.Slot, error)
+}
+
+func (sr *ServiceRepository) RetrieveAllSlots(page int, status bool) ([]models.Slot, error) {
+	var slots []models.Slot
+
+	query := sr.db.Offset(page * DEFAULT_PAGE_SIZE).Limit(DEFAULT_PAGE_SIZE)
+	if !status {
```
Comment (vraravam) [inline repositories/slot_repository.go]: simplify this by using `status` to generate an array of possible values, which in turn can be used in a `where` clause.

File: repositories/slot_repository.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+		query = query.Where("status IN (?, ?)", "Booked", "Reserved")
+	} else {
+		query = query.Where("status = ?", "Available")
+	}
+	if err := query.Find(&slots).Error; err != nil {
```
Comment (vraravam) [inline repositories/slot_repository.go]: assignment and conditionals in the same line (though supported in go) are confusing. please separate out into 2 lines

**Review by vraravam — CHANGES_REQUESTED** [review]

**Review by vraravam — APPROVED** [review]
